src/sentry/statistical_detectors/detector.py

```diff
@@ -55,6 +67,127 @@ def bulk_write_states(self, payloads: List[DetectorPayload], states: List[T]):
         ...
+
+    @property
+    @abstractmethod
+    def state(self) -> DetectorState:
+        ...
+
+
+@dataclass(frozen=True)
+class RegressionDetector(ABC):
+    kind: str
+    config: DetectorConfig
+    store: DetectorStore
+    state_cls: type[DetectorState]
+    detector_cls: type[DetectorAlgorithm]
+
+    @classmethod
+    def all_payloads(
+        cls,
+        projects: List[Project],
+        start: datetime,
+    ) -> Generator[DetectorPayload, None, None]:
+        projects_per_query = options.get("statistical_detectors.query.batch_size")
+        assert projects_per_query > 0
+
+        for projects in chunked(projects, projects_per_query):
+            try:
+                yield from cls.query_payloads(projects, start)
+            except Exception as e:
+                sentry_sdk.capture_exception(e)
+                continue
+
+    @classmethod
+    @abstractmethod
+    def query_payloads(
+        cls,
+        projects: List[Project],
+        start: datetime,
+    ) -> List[DetectorPayload]:
+        ...
+
+    @classmethod
+    def detect_trends(
+        cls, projects: List[Project], start: datetime
+    ) -> Generator[Tuple[Optional[TrendType], float, DetectorPayload], None, None]:
+        unique_project_ids: Set[int] = set()
+
+        total_count = 0
+        regressed_count = 0
+        improved_count = 0
+
+        for payloads in chunked(cls.all_payloads(projects, start), 100):
+            total_count += len(payloads)
+
+            raw_states = cls.store.bulk_read_states(payloads)
+
+            states = []
+
+            for raw_state, payload in zip(raw_states, payloads):
+                try:
+                    state = cls.state_cls.from_redis_dict(raw_state)
+                except Exception as e:
+                    state = cls.state_cls.empty()
+
+                    if raw_state:
+                        # empty raw state implies that there was no
+                        # previous state so no need to capture an exception
+                        sentry_sdk.capture_exception(e)
+
+                algorithm = cls.detector_cls(state, cls.config)
+                trend_type, score = algorithm.update(payload)
+
+                states.append(None if trend_type is None else algorithm.state.to_redis_dict())
+
+                if trend_type == TrendType.Regressed:
+                    regressed_count += 1
+                elif trend_type == TrendType.Improved:
+                    improved_count += 1
+
+                unique_project_ids.add(payload.project_id)
+
+                yield (trend_type, score, payload)
+
+            cls.store.bulk_write_states(payloads, states)
+
+        # TODO: These metrics are function specific. Refactor so that it handles
+        # various detector types.
+
+        # This is the total number of functions examined in this iteration
+        metrics.incr(
+            "statistical_detectors.total.functions",
```

Comment:
Looks like you could just f-string these metric names with the `kind` property?

I guess `statistical_detectors.profiling.projects.active` makes slightly less sense.

Or you could return the values as a tuple and handle them in an overridden `detect_trends` implementation?

Reply:
I will clean all this up in a follow-up to do roughly what you described. I just don't want to change the metrics and do the refactoring in the same PR right now